### PR TITLE
Feature/regroup mechanic

### PR DESCRIPTION
**Pull request by CheBato**
- Added tests and bug fix for the regroup phase.

### CODE REVIEW

**Comment by danbastin** [comment]
A couple of small tests to ensure that claim > pass and pass > pass also start the regroup phase. Do we also want to include tests for phase specific effects rolling off here such as chirruts ability or on defeats caused by health buffs rolling off?

**Comment by CheBato** [comment]
> A couple of small tests to ensure that claim > pass and pass > pass also start the regroup phase. Do we also want to include tests for phase specific effects rolling off here such as chirruts ability or on defeats caused by health buffs rolling off?

Thanks for the comment! 
I think the claim > pass and pass > pass tests are included in the initiative spec I figured it makes more sense having them there. This is a personal opinion but I think chirruts should be handled in its own spec while I agree for the defeats and stat modifiers rolling off that they should be included I'll add them.

**Comment by AMMayberry1** [comment]
> A couple of small tests to ensure that claim > pass and pass > pass also start the regroup phase. Do we also want to include tests for phase specific effects rolling off here such as chirruts ability or on defeats caused by health buffs rolling off?

**Comment by AMMayberry1** [comment]
> > A couple of small tests to ensure that claim > pass and pass > pass also start the regroup phase. Do we also want to include tests for phase specific effects rolling off here such as chirruts ability or on defeats caused by health buffs rolling off?
> 
> Thanks for the comment! I think the claim > pass and pass > pass tests are included in the initiative spec I figured it makes more sense having them there. This is a personal opinion but I think chirruts should be handled in its own spec while I agree for the defeats and stat modifiers rolling off that they should be included I'll add them.

This makes sense to me

File: server/game/core/gameSteps/phases/Phase.ts [hunk 1]
```diff
@@ -42,11 +42,11 @@ export abstract class Phase extends BaseStepWithPipeline {
...
     protected endPhase(skipEventWindow = false): void {
         if (!skipEventWindow) {
-            this.game.createEventAndOpenWindow(EventName.OnPhaseEnded, { phase: this.name }, TriggerHandlingMode.ResolvesTriggers);
+            this.game.createEventAndOpenWindow(EventName.OnPhaseEnded, { phase: this.name }, TriggerHandlingMode.ResolvesTriggers, () => this.game.currentPhase = null);
 
             // for post-phase state cleanup. emit directly, don't need a window.
             this.game.emit(EventName.OnPhaseEndedCleanup, { phase: this.name });
         }
-        this.game.currentPhase = null;
+        // this.game.currentPhase = null;
```
Comment (AMMayberry1) [inline server/game/core/gameSteps/phases/Phase.ts:50]: Remove

File: test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                    // old hand & deck setup
+                    const oldHandPlayer1 = [];
+                    const oldHandPlayer2 = [];
+                    const oldDeckPlayer1 = [];
+                    const oldDeckPlayer2 = [];
+                    const oldResourcesPlayer1 = [];
+                    const oldResourcesPlayer2 = [];
+
+                    // we need to create a deep copy since otherwise the original array receives changes
+                    context.player1.hand.forEach((val) => oldHandPlayer1.push(val));
```
Comment (AMMayberry1) [inline test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts:36]: You can consolidate each of these into one line with their definitions like this:

```typescript
const oldHandPlayer1 = [...context.player1.hand];
```

the `...` operator unpacks the array elements

File: test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                    // exhausted units need to be exhausted until the ready card phase
+                    expect(context.allianceXwing.exhausted).toBe(true);
+                    expect(context.wampa.exhausted).toBe(true);
+                    expect(context.tielnFighter.exhausted).toBe(true);
+
+                    // Resource a Card
+                    // Player 1 Resources a card and Player 2 doesn't
+                    context.player1.clickCard('wroshyr-tree-tender');
+                    // this is the index of wroshyr-tree-tender
+                    oldResourcesPlayer1.push(oldHandPlayer1[1]);
```
Comment (AMMayberry1) [inline test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts:72]: Can also do it as `oldResourcesPlayer1.push(context.wroshyrTreeTender);` for clarity

File: test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                            spaceArena: ['alliance-xwing'],
+                        },
+                        player2: {
+                            groundArena: ['wampa'],
+                        }
+                    });
+
+                    const { context } = contextRef;
+
+                    context.player1.clickCard(context.attackPatternDelta);
```
Comment (AMMayberry1) [inline test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts:122]: This is great! Couple additional things would be good:

- Can you add break up this block of code with some comments explaining what the clicks are doing, it's a little hard to follow
- It would also be good to explicitly test the timing on the defeat. General Krell is fantastic for that, could you add him for player 1 and confirm that when his ability prompt comes up that we are still in the action phase

Comment (CheBato) [reply test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts:122]: Great idea! I'll add him to the tests

File: test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts (hunk 3 above)
Comment (CheBato) [reply test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts:72]: yes sir!

File: test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts (hunk 2 above)
Comment (CheBato) [reply test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts:36]: great will use ... instead

File: server/game/core/gameSteps/phases/Phase.ts (hunk 1 above)
Comment (CheBato) [reply server/game/core/gameSteps/phases/Phase.ts:50]: hell yeah!

File: test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts [hunk 5]
```diff
@@ -119,9 +111,16 @@ describe('Regroup phase', function() {
...
+                    // Play card Attack pattern delta
                     context.player1.clickCard(context.attackPatternDelta);
+                    // Select Ardent Sympathizer to get +3/+3
                     context.player1.clickCard(context.ardentSympathizer);
+                    // Select Scout Bike Pursuer to get +2/+2
                     context.player1.clickCard(context.scoutBikePursuer);
+                    // Select General Krell to receive +1/+1
+                    context.player1.clickCard(context.generalKrell);
+
+                    // Select wampa to attack Ardent Sympathizer
```
Comment (AMMayberry1) [inline test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts:123]: nit: maybe note that this attack damages it enough to be defeated when the effect expires so the intent is clear

Comment (CheBato) [reply test/server/core/gameSteps/regroupWindow/RegroupPhase.spec.ts]: gotcha!

**Review by AMMayberry1 — APPROVED** [review]
